openpgp/keys.go

```diff
@@ -419,9 +419,17 @@ EachPacket:
 
 			// Next handle the case of a self-signature. According to RFC8440,
 			// Section 5.2.3.3, if there are several self-signatures,
-			// we should take the newer one.
+			// we should take the newer one.  If they were both created
+			// at the same time, but one of them has keyflags specified and the
+			// other doesn't, keep the one with the keyflags. We have actually
+			// seen this in the wild (see the 'Yield' test in read_test.go).
+			// If there is a tie, and both have the same value for FlagsValid,
+			// then "last writer wins."
 			if current != nil &&
-				(current.SelfSignature == nil || pkt.CreationTime.After(current.SelfSignature.CreationTime)) &&
+				(current.SelfSignature == nil ||
+					!pkt.CreationTime.Before(current.SelfSignature.CreationTime) ||
```

Comment:
Should this line go back to being `After` instead of `!Before`? I think you're accidentally handling the `Equal` case twice.